### PR TITLE
kubelet: logs: do not wait when following terminated container

### PR DESCRIPTION
**Pull request by sjenning**
Currently, a `kubectl logs -f` on a terminated container will output the logs, wait 5 seconds (`stateCheckPeriod`), then return.  The 5 seconds delay should not occur as the container is terminated and unable to generate additional log messages.

This PR puts a check at the beginning of `waitLogs()` to avoid doing the wait when the container is not running.

@derekwaynecarr @smarterclayton

### CODE REVIEW

**Comment by jayunit100** [comment]
Makes sense - by definition termination means that specific pod is never coming back.... do we need a corresponding issue for this PR?

File: pkg/kubelet/kuberuntime/logs/logs.go [hunk 1]
```diff
@@ -354,9 +354,28 @@ func ReadLogs(path, containerID string, opts *LogOptions, runtimeService interna
 	}
 }
 
+func isPodRunning(id string, r internalapi.RuntimeService) (bool, error) {
+	s, err := r.ContainerStatus(id)
+	if err != nil {
+		return false, err
+	}
+	// Only keep following container log when it is running.
+	if s.State != runtimeapi.ContainerState_CONTAINER_RUNNING {
```
Comment (feiskyer) [inline pkg/kubelet/kuberuntime/logs/logs.go]: The container may in starting state, use `s.State == runtimeapi.ContainerState_CONTAINER_EXITED` instead?

Comment (sjenning) [reply pkg/kubelet/kuberuntime/logs/logs.go]: I was just reusing the logic that already existed there as to not complicate things.

File: pkg/kubelet/kuberuntime/logs/logs.go [hunk 2]
```diff
@@ -354,9 +354,28 @@ func ReadLogs(path, containerID string, opts *LogOptions, runtimeService interna
 	}
 }
 
+func isPodRunning(id string, r internalapi.RuntimeService) (bool, error) {
```
[the remark this answers is not in the document]
Comment (sjenning) [reply pkg/kubelet/kuberuntime/logs/logs.go]: I assume you mean is*Container*Running.  And yes that would be the proper name :-)

**Comment by derekwaynecarr** [comment]
this is a cleanup and improvement on the existing behavior.

/approve
/lgtm
/retest

File: pkg/kubelet/kuberuntime/logs/logs.go [hunk 3]
```diff
@@ -354,9 +354,28 @@ func ReadLogs(path, containerID string, opts *LogOptions, runtimeService interna
 	}
 }
 
+func isContainerRunning(id string, r internalapi.RuntimeService) (bool, error) {
+	s, err := r.ContainerStatus(id)
```
[the remark this answers is not in the document]
Comment (sjenning) [reply pkg/kubelet/kuberuntime/logs/logs.go]: Just to be clear, this PR didn't introduce this behavior.  It just didn't fix it.

[the remark this answers is not in the document]
Comment (smarterclayton) [reply pkg/kubelet/kuberuntime/logs/logs.go]: Yes, I just also saw it while testing and observed it in the code.  This code can't safely assume that ContainerStatus can return a pod so we should fix it while we're in here.

[the remark this answers is not in the document]
Comment (sjenning) [reply pkg/kubelet/kuberuntime/logs/logs.go]: follow on PR for this https://github.com/kubernetes/kubernetes/pull/63334

**Pull request merged**